improvement(client-container-runtime): compat utils cleanup

jason-ha opened the pull request:
- correct compatibility range typo
- `*Compat[ibility]Mode` -> `*Compat[ibility]Version`
- make `runtimeOptionsAffectingDocSchemaConfigMap` immutable constant
- separate logical `getCompatibilityVersionDefaults` from testable unit `getConfigsForCompatMode`
- reflow wide comments and add links
- change `defaultCompatibilityVersion` to allow more natural config map for `explicitSchemaControl`
- correct typing for `getConfigsForCompatMode` to accurately note that only partial `T` can be guaranteed. This exposes potential hole in current logic and will need to be addressed. Compilation is preserved by using an inappropriate cast.
  - types in unit test for `getConfigsForCompatMode` are updated to be more realistic and then reveal the `Partial<T>` issue more directly.
- add test case for no runtime options sent to `ContainerRuntime.loadRuntime`

jason-ha committed:
improvement(client-container-runtime): compat utils cleanup
- `*Compat[ibility]Mode` -> `*Compat[ibility]Version`
- make `runtimeOptionsAffectingDocSchemaConfigMap` immutable constant
- separate logical `getCompatibilityVersionDefaults` from testable unit `getConfigsForCompatMode`
- reflow wide comments and add links
- change `defaultCompatibilityVersion` to allow more natural config map for `explicitSchemaControl`
- correct typing for `getConfigsForCompatMode` to accurately note that only partial `T` can be guaranteed. This exposes potential hole in current logic and will need to be addressed. Compilation is preserved by using an inappropriate cast.
  - types in unit test for `getConfigsForCompatMode` are updated to be more realistic and then reveal the `Partial<T>` issue more directly.`

scottn12 commented on packages/runtime/container-runtime/src/compatUtils.ts:
```diff
@@ -16,48 +16,68 @@ import {
+ * @remarks Only 2.0.0-defaults is expected, but index signatures cannot be a
+ * literal; so, just allow any major -defaults prerelease.
+ */
+export type MinimumMinorSemanticVersion = `${bigint}.${bigint}.0` | `${bigint}.0.0-defaults`;
```
I wonder if there may be scenarios where we do actually want to specify a patch version. What immediately comes to mind is if there is a major bug that gets fixed in a patch release. For example, a new feature that is added in say `2.50.0`, but needs to be patched. In that case, maybe we want to specify the version as `2.50.1`

jason-ha replied:
Going to patch by changing the default a customer gets when specifying 2.50.0 in a patch such that they need to specify 2.50.1?
I recommend putting a stake in the ground and relaxing this should there be a need. I don't currently expect this type will be exposed.

scottn12 replied:
I believe it likely will be exposed to customers, since they will be able to specify their `compatibilityMode`/`compatibilityVersion` in the future and the input type would be `MinimumMinorSemanticVersion`

jason-ha replied:
We are going to expose the table? This new type is explicitly for the table to limit the values that are entered to be patterns we expect. Anything else would need to be thought through.
I don't think this will be exposed. Customers can/will use `SemanticVersion`.

scottn12 replied:
Ah gotcha, I misunderstood. This should be okay then

scottn12 commented on packages/runtime/container-runtime/src/compatUtils.ts:
```diff
@@ -16,48 +16,68 @@ import {
+ * is semantically less than 2.0.0.
  */
-export const defaultCompatibilityMode = "2.0.0" as const;
+export const defaultCompatibilityVersion = "2.0.0-defaults" as const;
```
This is a topic I discussed with @anthony-murphy. I initially had something similar, but we came to the conclusion that having a special default version added more confusion for FF devs. That said, the way you implemented this is a bit less confusing than my implementation, so we may want to rethink this.

jason-ha replied:
If this still isn't appealing another option would be to encode `>` into the keys and modify the version check operator. OR use version ranges in keys (supported by semver).

scottn12 replied:
Personally, I'm okay with `2.0.0-defaults`. The main concern we had with my previous implementation is that we did a check for the hardcoded default version, then returned a hardcoded preset set of configurations for it. This implementation avoids that issue and uses the config map naturally

scottn12 reviewed (commented):
Thanks for doing this Jason, these are really great improvements! I think I'd like to hold off on changing `compatibilityMode` -> `compatibilityVersion` in this PR. I was planning on changing the name in a PR where we add it to the public API surface so we can have that discussion there.

jason-ha commented on packages/runtime/container-runtime/src/compatUtils.ts:
```diff
@@ -82,59 +102,78 @@ export type RuntimeOptionsAffectingDocSchema = Required<
+		runtimeOptionsAffectingDocSchemaConfigMap,
+		// This is a bad cast away from Partial that getConfigsForCompatMode provides.
+		// ConfigMap should be restructured to provide RuntimeOptionsAffectingDocSchema guarantee.
+	) as RuntimeOptionsAffectingDocSchema;
```
Thanks Copilot. Did you notice that I placed a comment explaining why we are doing this acknowledged bad cast?

jason-ha commented:
> Thanks for doing this Jason, these are really great improvements! I think I'd like to hold off on changing `compatibilityMode` -> `compatibilityVersion` in this PR. I was planning on changing the name in a PR where we add it to the public API surface so we can have that discussion there.

I made the rename because I needed distinguish places that were potentially `CompatibilityMode` (real API we have) from this. I had also done it locally before I read the other outstanding comment in the original PR. There were some renames that I missed (I noticed some in test cases). I am okay reverting, but I see value in being clear about version versus mode right now. Would it cause trouble in future plans to do this now? I will go with your preference.

scottn12 commented:
Yeah, it should be fine to leave the rename in. It doesn't really make a difference for our future plans

jason-ha committed:
style,test(client-container-runtime): more cleanup
- replace remaining `compatibilityMode` with `compatibilityVersion`
- add coverage for case when no runtime options are used
- comment why all undefined runtime options test case exists
- improve type workaround for `enableRuntimeIdCompressor`

scottn12 reviewed (approved):
Thanks for doing this! I really like all the syntax improvements :)

jason-ha committed:
docs(client-container-runtime): fix compatibility comment

Co-authored-by: Scott Norton <scottnorton@microsoft.com>
